File: Plugins/VanillaBasics/src/main/java/org/tobi29/scapes/vanilla/basics/VanillaBasicsCrafting.java
package org.tobi29.scapes.vanilla.basics;

import java8.util.stream.Collectors;
import org.tobi29.scapes.block.GameRegistry;
import org.tobi29.scapes.block.ItemStack;
import org.tobi29.scapes.engine.utils.Streams;
import org.tobi29.scapes.entity.client.MobPlayerClientMain;
import org.tobi29.scapes.entity.server.MobPlayerServer;
import org.tobi29.scapes.vanilla.basics.material.*;
import org.tobi29.scapes.vanilla.basics.util.ToolUtil;

import java.util.Arrays;
import java.util.List;

class VanillaBasicsCrafting {
    static void registerRecipes(VanillaBasics plugin, GameRegistry registry) {
        GameRegistry.Registry<CropType> cropRegistry =
                registry.get("VanillaBasics", "CropType");
        GameRegistry.Registry<TreeType> treeRegistry =
                registry.get("VanillaBasics", "TreeType");
        GameRegistry.Registry<StoneType> stoneRegistry =
                registry.get("VanillaBasics", "StoneType");
        registerRecipesBasics(plugin, registry, treeRegistry, stoneRegistry);
        registerRecipesStone(plugin, registry);
        registerRecipesFood(plugin, registry, cropRegistry, stoneRegistry);
        registerRecipesMetal(plugin, registry, stoneRegistry);
        registerRecipesIron(plugin, registry, treeRegistry);
    }

    static void registerRecipesBasics(VanillaBasics plugin,
            GameRegistry registry, GameRegistry.Registry<TreeType> treeRegistry,
            GameRegistry.Registry<StoneType> stoneRegistry) {
        VanillaMaterial materials = plugin.materials;
        CraftingRecipeType recipeType = new CraftingRecipeType() {
            @Override
            public String name() {
                return "Basics";
            }

            @Override
            public boolean table() {
                return false;
            }

            @Override
            public boolean availableFor(MobPlayerServer player) {
                return true;
            }

            @Override
            public boolean availableFor(MobPlayerClientMain player) {
                return true;
            }
        };
        List<ItemStack> hammer = Arrays.asList(materials.stoneHammer.example(1),
                materials.metalHammer.example(1));
        List<ItemStack> saw = Arrays.asList(materials.stoneSaw.example(1),
                materials.metalSaw.example(1));
        List<ItemStack> plank = Streams.of(treeRegistry.values())
                .map(tree -> new ItemStack(materials.wood, tree.data(registry)))
                .collect(Collectors.toList());

        plugin.c.craftingRecipe(recipeType, c -> {
            c.ingredients.add(plank);
            c.ingredients.add(hammer);
            c.requirements.add(saw);
            c.result = new ItemStack(materials.craftingTable, 0);
        }, "vanilla.basics.crafting.basics.CraftingTable");
        plugin.c.craftingRecipe(recipeType, c -> {
            c.ingredients.add(plank);
            c.requirements.add(saw);
            c.result = new ItemStack(materials.chest, 0);
        }, "vanilla.basics.crafting.basics.Chest");
        plugin.c.craftingRecipe(recipeType, c -> {
            c.ingredients.add(plank);
            c.ingredients.add(new ItemStack(materials.string, 1));
            c.requirements.add(saw);
            c.result = new ItemStack(materials.researchTable, 0);
        }, "vanilla.basics.crafting.basics.ResearchTable");
        Streams.of(stoneRegistry.values())
                .mapToInt(stoneType -> stoneType.data(registry))
                .forEach(data -> plugin.c.craftingRecipe(recipeType, c -> {
                    c.ingredients
                            .add(new ItemStack(materials.stoneRock, data, 9));
                    c.requirements.add(hammer);
                    c.result = new ItemStack(materials.cobblestone, data);
                }, "vanilla.basics.crafting.basics.Cobblestone" + data));
        plugin.c.craftingRecipe(recipeType, c -> {
            c.ingredients.add(new ItemStack(materials.stick, 0));
            c.ingredients.add(new ItemStack(materials.coal, 0));
            c.result = new ItemStack(materials.torch, 0);
        }, "vanilla.basics.crafting.basics.Torch");

        plugin.addCraftingRecipe(recipeType);
    }

    static void registerRecipesStone(VanillaBasics plugin,
            GameRegistry registry) {
        VanillaMaterial materials = plugin.materials;
        CraftingRecipeType recipeType = new CraftingRecipeType() {
            @Override
            public String name() {
                return "Stone";
            }

            @Override
            public boolean table() {
                return false;
            }

            @Override
            public boolean availableFor(MobPlayerServer player) {
                return true;
            }

            @Override
            public boolean availableFor(MobPlayerClientMain player) {
                return true;
            }
        };
        ItemStack flint = new ItemStack(materials.stoneRock,
                StoneType.FLINT.data(registry));
        ItemStack hoe = new ItemStack(flint);
        ToolUtil.createStoneTool(materials.plugin, hoe, "Hoe");
        ItemStack hoeHead = new ItemStack(hoe).setData(0);
        ItemStack hammer = new ItemStack(flint);
        ToolUtil.createStoneTool(materials.plugin, hammer, "Hammer");
        ItemStack hammerHead = new ItemStack(hammer).setData(0);
        ItemStack saw = new ItemStack(flint);
        ToolUtil.createStoneTool(materials.plugin, saw, "Saw");
        ItemStack sawHead = new ItemStack(saw).setData(0);
        ItemStack axe = new ItemStack(flint);
        ToolUtil.createStoneTool(materials.plugin, axe, "Axe");
        ItemStack axeHead = new ItemStack(axe).setData(0);
        ItemStack shovel = new ItemStack(flint);
        ToolUtil.createStoneTool(materials.plugin, shovel, "Shovel");
        ItemStack shovelHead = new ItemStack(shovel).setData(0);
        ItemStack pickaxe = new ItemStack(flint);
        ToolUtil.createStoneTool(materials.plugin, pickaxe, "Pickaxe");
        ItemStack pickaxeHead = new ItemStack(pickaxe).setData(0);
        ItemStack sword = new ItemStack(flint);
        ToolUtil.createStoneTool(materials.plugin, sword, "Sword");
        ItemStack swordHead = new ItemStack(sword).setData(0);

        plugin.c.craftingRecipe(recipeType, c -> {
            c.ingredients.add(flint);
            c.requirements.add(flint);
            c.result = hoeHead;
        }, "vanilla.basics.crafting.stone.HoeHead");
        plugin.c.craftingRecipe(recipeType, c -> {
            c.ingredients.add(flint);
            c.requirements.add(flint);
            c.result = hammerHead;
        }, "vanilla.basics.crafting.stone.HammerHead");
        plugin.c.craftingRecipe(recipeType, c -> {
            c.ingredients.add(flint);
            c.requirements.add(flint);
            c.result = sawHead;
        }, "vanilla.basics.crafting.stone.SawHead");
        plugin.c.craftingRecipe(recipeType, c -> {
            c.ingredients.add(flint);
            c.requirements.add(flint);
            c.result = axeHead;
        }, "vanilla.basics.crafting.stone.AxeHead");
        plugin.c.craftingRecipe(recipeType, c -> {
            c.ingredients.add(flint);
            c.requirements.add(flint);
            c.result = shovelHead;
        }, "vanilla.basics.crafting.stone.ShovelHead");
        plugin.c.craftingRecipe(recipeType, c -> {
            c.ingredients.add(flint);
            c.requirements.add(flint);
            c.result = pickaxeHead;
        }, "vanilla.basics.crafting.stone.PickaxeHead");
        plugin.c.craftingRecipe(recipeType, c -> {
            c.ingredients.add(flint);
            c.requirements.add(flint);
            c.result = swordHead;
        }, "vanilla.basics.crafting.stone.SwordHead");
        plugin.c.craftingRecipe(recipeType, c -> {
            c.ingredients.add(hoe);
            c.ingredients.add(hoeHead);
            c.ingredients.add(hammer);
            c.ingredients.add(hammerHead);
            c.ingredients.add(saw);
            c.ingredients.add(sawHead);
            c.ingredients.add(axe);
            c.ingredients.add(axeHead);
            c.ingredients.add(shovel);
            c.ingredients.add(shovelHead);
            c.ingredients.add(pickaxe);
            c.ingredients.add(pickaxeHead);
            c.ingredients.add(sword);
            c.ingredients.add(swordHead);
            c.result = flint;
        }, "vanilla.basics.crafting.stone.BreakTool");
        plugin.c.craftingRecipe(recipeType, c -> {
            c.ingredients.add(new ItemStack(materials.grassBundle, 0, 2));
            c.result = new ItemStack(materials.string, 0);
        }, "vanilla.basics.crafting.stone.String");
        plugin.c.craftingRecipe(recipeType, c -> {
            c.ingredients.add(new ItemStack(materials.string, 0, 8));
            c.result = new ItemStack(materials.string, 1);
        }, "vanilla.basics.crafting.stone.Fabric");
        plugin.c.craftingRecipe(recipeType, c -> {
            c.ingredients.add(new ItemStack(materials.grassBundle, 0, 2));
            c.result = new ItemStack(materials.straw, 0);
        }, "vanilla.basics.crafting.stone.WetStraw");
        plugin.c.craftingRecipe(recipeType, c -> {
            c.ingredients.add(new ItemStack(materials.grassBundle, 1, 2));
            c.result = new ItemStack(materials.straw, 1);
        }, "vanilla.basics.crafting.stone.Straw");

        plugin.addCraftingRecipe(recipeType);
    }

    static void registerRecipesFood(VanillaBasics plugin, GameRegistry registry,
            GameRegistry.Registry<CropType> cropRegistry,
            GameRegistry.Registry<StoneType> stoneRegistry) {
        VanillaMaterial materials = plugin.materials;
        CraftingRecipeType recipeType = new CraftingRecipeType() {
            @Override
            public String name() {
                return "Food";
            }

            @Override
            public boolean table() {
                return true;
            }

            @Override
            public boolean availableFor(MobPlayerServer player) {
                return player.metaData("Vanilla").getStructure("Research")
                        .getStructure("Finished").getBoolean("Food");
            }

            @Override
            public boolean availableFor(MobPlayerClientMain player) {
                return player.metaData("Vanilla").getStructure("Research")
                        .getStructure("Finished").getBoolean("Food");
            }
        };
        List<ItemStack> cobblestones = Streams.of(stoneRegistry.values())
                .filter(stoneType -> stoneType.resistance() > 0.1)
                .mapToInt(stoneType -> stoneType.data(registry))
                .mapToObj(data -> new ItemStack(materials.cobblestone, data, 2))
                .collect(Collectors.toList());
        List<ItemStack> pickaxe =
                Arrays.asList(materials.stonePickaxe.example(1),
                        materials.metalPickaxe.example(1));
        List<ItemStack> hammer = Arrays.asList(materials.stoneHammer.example(1),
                materials.metalHammer.example(1));

        // TODO: Replace with oven
        plugin.c.craftingRecipe(recipeType, c -> {
            c.ingredients.add(cobblestones);
            c.ingredients.add(new ItemStack(materials.stick, 0, 4));
            c.requirements.add(pickaxe);
            c.result = new ItemStack(materials.furnace, 0);
        }, "vanilla.basics.crafting.food.Furnace");
        plugin.c.craftingRecipe(recipeType, c -> {
            c.ingredients.add(cobblestones);
            c.requirements.add(hammer);
            c.result = new ItemStack(materials.quern, 0);
        }, "vanilla.basics.crafting.food.Quern");
        Streams.of(cropRegistry.values())
                .mapToInt(cropType -> cropType.data(registry))
                .forEach(data -> plugin.c.craftingRecipe(recipeType, c -> {
                    c.ingredients.add(new ItemStack(materials.grain, data, 8));
                    c.result = new ItemStack(materials.dough, data);
                }, "vanilla.basics.crafting.food.Dough" + data));
        plugin.c.craftingRecipe(recipeType, c -> {
            c.ingredients.add(cobblestones);
            c.requirements.add(hammer);
            c.result = new ItemStack(materials.quern, 0);
        }, "vanilla.basics.crafting.");

        plugin.addCraftingRecipe(recipeType);
    }

    static void registerRecipesMetal(VanillaBasics plugin,
            GameRegistry registry,
            GameRegistry.Registry<StoneType> stoneRegistry) {
        VanillaMaterial materials = plugin.materials;
        CraftingRecipeType recipeType = new CraftingRecipeType() {
            @Override
            public String name() {
                return "Metal";
            }

            @Override
            public boolean table() {
                return true;
            }

            @Override
            public boolean availableFor(MobPlayerServer player) {
                return player.metaData("Vanilla").getStructure("Research")
                        .getStructure("Finished").getBoolean("Metal");
            }

            @Override
            public boolean availableFor(MobPlayerClientMain player) {
                return player.metaData("Vanilla").getStructure("Research")
                        .getStructure("Finished").getBoolean("Metal");
            }
        };
        List<ItemStack> cobblestones = Streams.of(stoneRegistry.values())
                .filter(stoneType -> stoneType.resistance() > 0.1)
                .mapToInt(stoneType -> stoneType.data(registry))
                .mapToObj(data -> new ItemStack(materials.cobblestone, data, 2))
                .collect(Collectors.toList());

        plugin.c.craftingRecipe(recipeType, c -> {
            c.ingredients.add(new ItemStack(materials.sand, 2));
            c.result = new ItemStack(materials.mold, 0);
        }, "vanilla.basics.crafting.metal.Mold");
        plugin.c.craftingRecipe(recipeType, c -> {
            c.ingredients.add(new ItemStack(materials.ingot, 0, 5));
            c.result = new ItemStack(materials.anvil, 0);
        }, "vanilla.basics.crafting.metal.Anvil");
        plugin.c.craftingRecipe(recipeType, c -> {
            c.ingredients.add(new ItemStack(materials.coal, 0, 8));
            c.result = new ItemStack(materials.forge, 0);
        }, "vanilla.basics.crafting.metal.Forge");
        plugin.c.craftingRecipe(recipeType, c -> {
            c.ingredients.add(cobblestones);
            c.requirements.add(new ItemStack(materials.metalPickaxe, 1));
            c.result = new ItemStack(materials.alloy, 0);
        }, "vanilla.basics.crafting.metal.Alloy");

        plugin.addCraftingRecipe(recipeType);
    }

    static void registerRecipesIron(VanillaBasics plugin, GameRegistry registry,
            GameRegistry.Registry<TreeType> treeRegistry) {
        VanillaMaterial materials = plugin.materials;
        CraftingRecipeType recipeType = new CraftingRecipeType() {
            @Override
            public String name() {
                return "Iron";
            }

            @Override
            public boolean table() {
                return true;
            }

            @Override
            public boolean availableFor(MobPlayerServer player) {
                return player.metaData("Vanilla").getStructure("Research")
                        .getStructure("Finished").getBoolean("Iron");
            }

            @Override
            public boolean availableFor(MobPlayerClientMain player) {
                return player.metaData("Vanilla").getStructure("Research")
                        .getStructure("Finished").getBoolean("Iron");
            }
        };
        List<ItemStack> plank = Streams.of(treeRegistry.values())
                .map(tree -> new ItemStack(materials.wood, tree.data(registry)))
                .collect(Collectors.toList());

        plugin.c.craftingRecipe(recipeType, c -> {
            c.ingredients.add(new ItemStack(materials.sand, 2));
            c.result = new ItemStack(materials.bloomery, 0);
        }, "vanilla.basics.crafting.iron.Bloomery");
        plugin.c.craftingRecipe(recipeType, c -> {
            c.ingredients.add(plank);
            c.ingredients.add(new ItemStack(materials.string, 1, 4));
            c.requirements.add(materials.metalSaw.example(1));
            c.result = new ItemStack(materials.bellows, 0);
        }, "vanilla.basics.crafting.iron.Bellows");

        plugin.addCraftingRecipe(recipeType);
    }
}

File: Plugins/VanillaBasics/VanillaBasicsCore/src/main/kotlin/org/tobi29/scapes/vanilla/basics/VanillaBasicsCrafting.kt
/*
 * Copyright 2012-2017 Tobi29
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.tobi29.scapes.vanilla.basics

import org.tobi29.scapes.block.GameRegistry
import org.tobi29.scapes.block.ItemStack
import org.tobi29.scapes.engine.utils.io.tag.map
import org.tobi29.scapes.engine.utils.io.tag.toBoolean
import org.tobi29.scapes.entity.client.MobPlayerClientMain
import org.tobi29.scapes.entity.server.MobPlayerServer
import org.tobi29.scapes.vanilla.basics.material.CraftingRecipeType
import org.tobi29.scapes.vanilla.basics.material.CropType
import org.tobi29.scapes.vanilla.basics.material.StoneType
import org.tobi29.scapes.vanilla.basics.material.TreeType
import org.tobi29.scapes.vanilla.basics.util.createStoneTool

internal fun VanillaBasics.registerRecipes(registry: GameRegistry) {
    val cropRegistry = registry.get<CropType>("VanillaBasics", "CropType")
    val treeRegistry = registry.get<TreeType>("VanillaBasics", "TreeType")
    val stoneRegistry = registry.get<StoneType>("VanillaBasics", "StoneType")
    registerRecipesBasics(registry, treeRegistry, stoneRegistry)
    registerRecipesStone(registry)
    registerRecipesFood(registry, cropRegistry, stoneRegistry)
    registerRecipesMetal(registry, stoneRegistry)
    registerRecipesIron(registry, treeRegistry)
}

private fun VanillaBasics.registerRecipesBasics(registry: GameRegistry,
                                                treeRegistry: GameRegistry.Registry<TreeType>,
                                                stoneRegistry: GameRegistry.Registry<StoneType>) {
    val recipeType = object : CraftingRecipeType() {
        override fun name(): String {
            return "Basics"
        }

        override fun table(): Boolean {
            return false
        }

        override fun availableFor(player: MobPlayerServer): Boolean {
            return true
        }

        override fun availableFor(player: MobPlayerClientMain): Boolean {
            return true
        }
    }

    val hammer = listOf(materials.flintHammer.example(1),
            materials.metalHammer.example(1))
    val saw = listOf(materials.flintSaw.example(1),
            materials.metalSaw.example(1))
    val plank = treeRegistry.values().asSequence().filterNotNull().map {
        ItemStack(materials.wood, it.data(registry))
    }.toList()

    crafting(recipeType) {
        recipe("vanilla.basics.crafting.basics.CraftingTable") {
            ingredients.add(plank)
            ingredients.add(hammer)
            requirements.add(saw)
            result = ItemStack(materials.craftingTable, 0)
        }
        recipe("vanilla.basics.crafting.basics.Chest") {
            ingredients.add(plank)
            requirements.add(saw)
            result = ItemStack(materials.chest, 0)
        }
        recipe("vanilla.basics.crafting.basics.ResearchTable") {
            ingredients.add(plank)
            ingredients.add(ItemStack(materials.string, 1))
            requirements.add(saw)
            result = ItemStack(materials.researchTable, 0)
        }
        stoneRegistry.values().asSequence().filterNotNull()
                .map { it.data(registry) }.forEach {
            recipe("vanilla.basics.crafting.basics.Cobblestone" + it, {
                ingredients.add(ItemStack(materials.stoneRock, it, 9))
                requirements.add(hammer)
                result = ItemStack(materials.cobblestone, it)
            })
        }
        recipe("vanilla.basics.crafting.basics.Torch") {
            ingredients.add(ItemStack(materials.stick, 0))
            ingredients.add(ItemStack(materials.coal, 0))
            result = ItemStack(materials.torch, 0)
        }
    }
}

private fun VanillaBasics.registerRecipesStone(registry: GameRegistry) {
    val recipeType = object : CraftingRecipeType() {
        override fun name(): String {
            return "Stone"
        }

        override fun table(): Boolean {
            return false
        }

        override fun availableFor(player: MobPlayerServer): Boolean {
            return true
        }

        override fun availableFor(player: MobPlayerClientMain): Boolean {
            return true
        }
    }

    val flint = ItemStack(materials.stoneRock,
            StoneType.FLINT.data(registry))
    val hoe = ItemStack(flint)
    createStoneTool(this, hoe, "Hoe")
    val hoeHead = ItemStack(hoe).setData(0)
    val hammer = ItemStack(flint)
    createStoneTool(this, hammer, "Hammer")
    val hammerHead = ItemStack(hammer).setData(0)
    val saw = ItemStack(flint)
    createStoneTool(this, saw, "Saw")
    val sawHead = ItemStack(saw).setData(0)
    val axe = ItemStack(flint)
    createStoneTool(this, axe, "Axe")
    val axeHead = ItemStack(axe).setData(0)
    val shovel = ItemStack(flint)
    createStoneTool(this, shovel, "Shovel")
    val shovelHead = ItemStack(shovel).setData(0)
    val pickaxe = ItemStack(flint)
    createStoneTool(this, pickaxe, "Pickaxe")
    val pickaxeHead = ItemStack(pickaxe).setData(0)
    val sword = ItemStack(flint)
    createStoneTool(this, sword, "Sword")
    val swordHead = ItemStack(sword).setData(0)

    crafting(recipeType) {
        recipe("vanilla.basics.crafting.stone.HoeHead") {
            ingredients.add(flint)
            requirements.add(flint)
            result = hoeHead
        }
        recipe("vanilla.basics.crafting.stone.HammerHead") {
            ingredients.add(flint)
            requirements.add(flint)
            result = hammerHead
        }
        recipe("vanilla.basics.crafting.stone.SawHead") {
            ingredients.add(flint)
            requirements.add(flint)
            result = sawHead
        }
        recipe("vanilla.basics.crafting.stone.AxeHead") {
            ingredients.add(flint)
            requirements.add(flint)
            result = axeHead
        }
        recipe("vanilla.basics.crafting.stone.ShovelHead") {
            ingredients.add(flint)
            requirements.add(flint)
            result = shovelHead
        }
        recipe("vanilla.basics.crafting.stone.PickaxeHead") {
            ingredients.add(flint)
            requirements.add(flint)
            result = pickaxeHead
        }
        recipe("vanilla.basics.crafting.stone.SwordHead") {
            ingredients.add(flint)
            requirements.add(flint)
            result = swordHead
        }
        recipe("vanilla.basics.crafting.stone.BreakTool") {
            ingredients.add(listOf(hoe, hoeHead, hammer, hammerHead, saw,
                    sawHead, axe,
                    axeHead, shovel, shovelHead, pickaxe, pickaxeHead,
                    sword, swordHead))
            result = flint
        }
        recipe("vanilla.basics.crafting.stone.String") {
            ingredients.add(ItemStack(materials.grassBundle, 0, 2))
            result = ItemStack(materials.string, 0)
        }
        recipe("vanilla.basics.crafting.stone.Fabric") {
            ingredients.add(ItemStack(materials.string, 0, 8))
            result = ItemStack(materials.string, 1)
        }
        recipe("vanilla.basics.crafting.stone.WetStraw") {
            ingredients.add(ItemStack(materials.grassBundle, 0, 2))
            result = ItemStack(materials.straw, 0)
        }
        recipe("vanilla.basics.crafting.stone.Straw") {
            ingredients.add(ItemStack(materials.grassBundle, 1, 2))
            result = ItemStack(materials.straw, 1)
        }
    }
}

private fun VanillaBasics.registerRecipesFood(registry: GameRegistry,
                                              cropRegistry: GameRegistry.Registry<CropType>,
                                              stoneRegistry: GameRegistry.Registry<StoneType>) {
    val recipeType = object : CraftingRecipeType() {
        override fun name(): String {
            return "Food"
        }

        override fun table(): Boolean {
            return true
        }

        override fun availableFor(player: MobPlayerServer): Boolean {
            return player.metaData("Vanilla").map("Research")?.map(
                    "Finished")?.get("Food")?.toBoolean() ?: false
        }

        override fun availableFor(player: MobPlayerClientMain): Boolean {
            return player.metaData("Vanilla").map("Research")?.map(
                    "Finished")?.get("Food")?.toBoolean() ?: false
        }
    }

    val cobblestones = stoneRegistry.values().asSequence().filterNotNull()
            .filter { it.resistance() > 0.1 }.map { it.data(registry) }
            .map { ItemStack(materials.cobblestone, it, 2) }.toList()
    val pickaxe = listOf(materials.flintPickaxe.example(1),
            materials.metalPickaxe.example(1))
    val hammer = listOf(materials.flintHammer.example(1),
            materials.metalHammer.example(1))

    crafting(recipeType) {
        // TODO: Replace with oven
        recipe("vanilla.basics.crafting.food.Furnace") {
            ingredients.add(cobblestones)
            ingredients.add(ItemStack(materials.stick, 0, 4))
            requirements.add(pickaxe)
            result = ItemStack(materials.furnace, 0)
        }
        recipe("vanilla.basics.crafting.food.Quern") {
            ingredients.add(cobblestones)
            requirements.add(hammer)
            result = ItemStack(materials.quern, 0)
        }
        cropRegistry.values().asSequence().filterNotNull()
                .map { it.data(registry) }.forEach {
            recipe("vanilla.basics.crafting.food.Dough" + it, {
                ingredients.add(ItemStack(materials.grain, it, 8))
                result = ItemStack(materials.dough, it)
            })
        }
        recipe("vanilla.basics.crafting.") {
            ingredients.add(cobblestones)
            requirements.add(hammer)
            result = ItemStack(materials.quern, 0)
        }
    }
}

private fun VanillaBasics.registerRecipesMetal(registry: GameRegistry,
                                               stoneRegistry: GameRegistry.Registry<StoneType>) {
    val recipeType = object : CraftingRecipeType() {
        override fun name(): String {
            return "Metal"
        }

        override fun table(): Boolean {
            return true
        }

        override fun availableFor(player: MobPlayerServer): Boolean {
            return player.metaData("Vanilla").map("Research")?.map(
                    "Finished")?.get("Metal")?.toBoolean() ?: false
        }

        override fun availableFor(player: MobPlayerClientMain): Boolean {
            return player.metaData("Vanilla").map("Research")?.map(
                    "Finished")?.get("Metal")?.toBoolean() ?: false
        }
    }
    val cobblestones = stoneRegistry.values().asSequence().filterNotNull()
            .filter { it.resistance() > 0.1 }.map { it.data(registry) }
            .map { ItemStack(materials.cobblestone, it, 2) }.toList()

    crafting(recipeType) {
        recipe("vanilla.basics.crafting.metal.Mold") {
            ingredients.add(ItemStack(materials.sand, 2))
            result = ItemStack(materials.mold, 0)
        }
        recipe("vanilla.basics.crafting.metal.Anvil") {
            ingredients.add(ItemStack(materials.ingot, 0, 5))
            result = ItemStack(materials.anvil, 0)
        }
        recipe("vanilla.basics.crafting.metal.Forge") {
            ingredients.add(ItemStack(materials.coal, 0, 8))
            result = ItemStack(materials.forge, 0)
        }
        recipe("vanilla.basics.crafting.metal.Alloy") {
            ingredients.add(cobblestones)
            requirements.add(ItemStack(materials.metalPickaxe, 1))
            result = ItemStack(materials.alloy, 0)
        }
    }
}

private fun VanillaBasics.registerRecipesIron(registry: GameRegistry,
                                              treeRegistry: GameRegistry.Registry<TreeType>) {
    val recipeType = object : CraftingRecipeType() {
        override fun name(): String {
            return "Iron"
        }

        override fun table(): Boolean {
            return true
        }

        override fun availableFor(player: MobPlayerServer): Boolean {
            return player.metaData("Vanilla").map("Research")?.map(
                    "Finished")?.get("Iron")?.toBoolean() ?: false
        }

        override fun availableFor(player: MobPlayerClientMain): Boolean {
            return player.metaData("Vanilla").map("Research")?.map(
                    "Finished")?.get("Iron")?.toBoolean() ?: false
        }
    }

    val plank = treeRegistry.values().asSequence().filterNotNull()
            .map { ItemStack(materials.wood, it.data(registry)) }
            .toList()

    crafting(recipeType) {
        recipe("vanilla.basics.crafting.iron.Bloomery") {
            ingredients.add(ItemStack(materials.sand, 2))
            result = ItemStack(materials.bloomery, 0)
        }
        recipe("vanilla.basics.crafting.iron.Bellows") {
            ingredients.add(plank)
            ingredients.add(ItemStack(materials.string, 1, 4))
            requirements.add(materials.metalSaw.example(1))
            result = ItemStack(materials.bellows, 0)
        }
    }
}
